### ISSUE TITLE
cs does not install scala runner for scala 3

### ISSUE THREAD

**Issue by bjornregnell** [coursier/coursier#2058]
See https://github.com/lampepfl/dotty/issues/12811

# Minimized example

Follow the official document Scala 3 Book, run the simple "Hello, world" example through scala3-repl.

```
   In the file Hello.scala, write

    @main def hello = println("Hello, world!")

    scala3-compiler Hello.scala
    scala3-repl hello
```

The scala3-repl was installed through Coursier, as suggested by the official webpage Download Scala 3

# Output

NO output.

The scala3-repl get into the interactive mode directly without any output.

# Expectation

A runner for scala3 should be installed, e.g.
```
scala3 hello
```
should produce `"Hello, world!"` as output, the same as in Scala 2.

**Comment by bjornregnell** [comment]
This is a showstopper for me to use cs in my teaching. And there is no .msi for Scala 3 so I'm more or less stuck (manually unpackaging a zip and going into environment variables to fiddle with path etc in Windows is not very attractive beginner programmer user experience...)

**Comment by bjornregnell** [comment]
See also https://github.com/lampepfl/dotty/issues/12502

**Comment by jacobus** [comment]
This issue is really a problem. Coursier is used in instructions in both the Scala Docs and the Dotty Installs and other places. Yet, if you follow the current instructions, then you end up with Scala 2.13. Can't find a nice and elegant way to just compile and run Scala 3 applications from the command line.

**Comment by bjornregnell** [comment]
Agree. The workaround I currently use is to download the zip files from the github release page and unpack somewhere locally,  and the fix your own path manually; I actually `ln -s` to each binary from my `~/bin` (which is on my path) to the scala3/bin but that's on Linux; less easy on Windows. This is not even a good experience for professionals but for beginner students it's a showstopper. This is not just a coursier problem but a problem for the Scala distribution as a whole.

**Comment by jacobus** [comment]
Thanks @bjornregnell . Is there anyone from the Coursier community who can help with this please. We just want to run a Scala 3 application after installing Scala 3 with Coursier. Ideally, it would be great to select which version of Scala should be used when using the scalac and scala commands.

**Comment by lrytz** [comment]
The `scala` coursier app (for Scala 2) works for both starting the REPL and launching a program (and also does scripts). This is done through the [`MainGenericRunner`](https://github.com/scala/scala/blob/2.13.x/src/repl-frontend/scala/tools/nsc/MainGenericRunner.scala).

**Comment by bjornregnell** [comment]
I think a `--launch x` by cs should be the **real** x and called by its name, and version differences should be set bya a `--version` option. I mean cs should delegate to the *actual* binaries called e.g. `scala`, `scalac`, `scalap`, `sbt`, etc., in the respective `bin` distribution. Otherwise people will get unpleasant surprises at some point and we contribute unnecessarily to confusion.

See also https://github.com/scalacenter/advisoryboard/pull/75

**Comment by jacobus** [comment]
Both https://www.scala-lang.org/download/scala3.html and http://dotty.epfl.ch/ recommends using coursier to get set up with Scala 3. Yet neither says how to run a Scala app. We are a month into the release of Scala 3, yet running hello world requires opening issues and asking questions on StackOverflow. Can this be so complex? Coursier lets you switch JDK versions, but you can't switch Scala versions. Am I missing something?

**Comment by bjornregnell** [comment]
@jacobus Agree! Being able to both launch and switch versions for the scala-runner is a "hygiene factor" and really important. 

One "easy" way to fix the runner issue is to provide "normal" install packages for each platform as we are used to for Scala 2, and it seems as it is about "just" switching CI tooling to use https://github.com/sbt/sbt-native-packager, see discussion at   https://github.com/lampepfl/dotty/issues/12502 and @anatoliykmetyuk is assigned to that issue.

If `cs` is "easy" to fix to make it (1) run Scala 3 apps and to (2) switch among versions, I don't know, but I definitely think it is high priority in that priority order. This relates to the general issue of making on-boarding of new-comers to Scala as smooth as possible. See related thread on this general matter here: https://contributors.scala-lang.org/t/next-steps-in-further-improving-community-experience-after-the-awesome-scala-3-release/5141/5

> Can this be so complex? [...] Am I missing something?

I don't know the technical details of coursier so I cannot tell how much work this would be. Someone else here that can make a "guestimate" of effort or explain "impossibilities" of the current architecture?

**Comment by jacobus** [comment]
Thanks for all the info @bjornregnell 
I've gone as far as[ asking on Stackoverflow](https://stackoverflow.com/questions/68041159/how-to-run-scala-3-applications-in-the-command-line-with-coursier) and no answer there either. Yet, both the Dotty site and Scala Lang site suggests using coursier as a way to get going with Scala 3.

**Comment by bjornregnell** [comment]
Thanks @jacobus  And thank *you* for engaging in this! Feel free to also chip into the Discord general thread on next steps in community experience that I linked above ;) :smiling_face_with_three_hearts: I think we all can help by providing our best input to priorities for improving on-boarding of new-comers to Scala. And we all know that there is *so* much still to do for this awesome open source community, hence the need to carefully consider priorities... (not easy at all, of course)

**Comment by bjornregnell** [comment]
@jacobus I just answered your [SO question](https://stackoverflow.com/a/68098274/10088022). 
HTH :)

**Comment by lrytz** [comment]
another workaround is `java -cp $(cs fetch -p scala3-repl):. myMain`

**Comment by jacobus** [comment]
Thanks gents, much appreciated. Those are both good workarounds for now. I still believe this is an issue though. Just imagine how confusing this is for someone new to Scala. I've been using Scala a great deal a long time ago and just wanted to play around with Scala 3 to see how things have changed. I could easily get things going with SBT, but thought I'll read the docs for a change, and they lead me to coursier, but I still don't know how to run a Scala 3 application with coursier from the command line. It's a bit nutty I must say.

**Comment by bjornregnell** [comment]
@lrytz That's a cool workaround! Do you want to add it to the SO question yourself, or do you want me to add it in an edit of my answer there?

**Comment by lrytz** [comment]
Just update your answer, no credits please :)

**Comment by bjornregnell** [comment]
Done!

**Comment by bjornregnell** [comment]
Wow, that's a list of dependencies just to run my little hello :)
```
$ echo "$(cs fetch -p scala3-repl)"
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-compiler_3/3.0.0/scala3-compiler_3-3.0.0.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-interfaces/3.0.0/scala3-interfaces-3.0.0.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.0.0/scala3-library_3-3.0.0.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/tasty-core_3/3.0.0/tasty-core_3-3.0.0.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-asm/9.1.0-scala-1/scala-asm-9.1.0-scala-1.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-sbt/compiler-interface/1.3.5/compiler-interface-1.3.5.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/org/jline/jline-reader/3.19.0/jline-reader-3.19.0.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/org/jline/jline-terminal/3.19.0/jline-terminal-3.19.0.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/org/jline/jline-terminal-jna/3.19.0/jline-terminal-jna-3.19.0.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.5/scala-library-2.13.5.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/com/google/protobuf/protobuf-java/3.7.0/protobuf-java-3.7.0.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-sbt/util-interface/1.3.0/util-interface-1.3.0.jar:
/home/bjornr/.cache/coursier/v1/https/repo1.maven.org/maven2/net/java/dev/jna/jna/5.3.1/jna-5.3.1.jar
```
Happy not having to type that myself :)

**Comment by lrytz** [comment]
Actually, `cs fetch -p org.scala-lang:scala3-library_3:3.0.0` would be enough for running an app (`scala3-library` and `scala-library`), the rest is for the REPL, which needs the compiler. Not if there's a way to simplify this to avoid the explicit versions `_3:3.0.0`.

**Comment by bjornregnell** [comment]
Aha. Many thanks! So for the record:
```
java -cp $(cs fetch -p org.scala-lang:scala3-library_3:3.0.0):. myMain
```

**Comment by xhudik** [comment]
this is horrible - if a newbie is supposed to use  `java -cp $(cs fetch -p org.scala-lang:scala3-library_3:3.0.0):. myMain` it is like [brainfuck](https://en.wikipedia.org/wiki/Brainfuck#Hello_World!) - no one will even try

**Comment by lrytz** [comment]
@xhudik we all agree that the current situation is not ideal, this is why the this issue exists. The workarounds are just that, workarounds that people can use in the meantime. Please keep comments constructive.

**Comment by SethTisue** [comment]
https://github.com/lampepfl/dotty/pull/13081 represents good progress towards this

with that PR merged and available in Scala 3 nightly builds, one can now:

	cs launch scala3-compiler:3+ -- hello.scala
    cs launch scala3-repl:3+ -M dotty.tools.MainGenericRunner -- hello

note the "3+" which gets you latest nightly.

note that `MainGenericRunner` can also run a script in a single step, rather than separate compile and run steps:

    cs launch scala3-repl:3+ -M dotty.tools.MainGenericRunner -- hello.scala

your script can even have dependencies:

    % cat parallel.scala
    import scala.collection.parallel.CollectionConverters._
    @main def hello =
      (1 to 5).par.foreach(println)
    % cs launch scala3-repl:3+ -M dotty.tools.MainGenericRunner \
    > --extra-jars $(cs fetch --intransitive org.scala-lang.modules:scala-parallel-collections_3:1.0.3) \
    > -- parallel.scala                                          
    1
    2
    5
    3
    4

**Comment by dwijnand** [comment]
Is `--intransitive` just being cautious or what happens without it?

**Comment by SethTisue** [comment]
> Is --intransitive just being cautious or what happens without it?

Without it you get multiple lines of output because both the Scala 2 and 3 libraries included in the output. (And the library versions might not match what you're actually using.)

**Comment by dwijnand** [comment]
> I didn't dig into whether `--extra-jars` can handle multiple lines of output

Ah, didn't notice earlier: you can use `-p` to get a "classpath-ready" string:

```
$ cs fetch org.scala-lang.modules:scala-parallel-collections_3:1.0.3
/Users/dnw/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-parallel-collections_3/1.0.3/scala-parallel-collections_3-1.0.3.jar
/Users/dnw/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.0.0/scala3-library_3-3.0.0.jar
/Users/dnw/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.5/scala-library-2.13.5.jar

$ cs fetch -p org.scala-lang.modules:scala-parallel-collections_3:1.0.3
/Users/dnw/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-parallel-collections_3/1.0.3/scala-parallel-collections_3-1.0.3.jar:/Users/dnw/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.0.0/scala3-library_3-3.0.0.jar:/Users/dnw/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.5/scala-library-2.13.5.jar
```

> whether it has any defense against ending up with multiple versions of the library JARs on the classpath.

At least in terms of classpath resources (/library.properties), it looks good:

```
$ echo '@main def Main = println(scala.util.Properties.versionString)' > foo.scala
$ cs launch scala3-compiler:3+ -M dotty.tools.MainGenericRunner -- foo.scala
version 2.13.6
$ cs launch scala3-compiler:3+ -M dotty.tools.MainGenericRunner --extra-jars $(cs fetch -p org.scala-lang:scala-library:2.13.5) -- foo.scala
version 2.13.6
```

**Comment by bjornregnell** [comment]
Does anyone know how to translate the workaround `java -cp $(cs fetch -p org.scala-lang:scala3-library_3:3.0.0):. myMain` to Windows cmd? It does not have to be a single line, just something that works at all... (200+ students coming soon and 80% Windows)

**Comment by dwijnand** [comment]
Looking at https://superuser.com/q/1137190/109200 it looks like it might be something like:

```cmd
for /F "tokens=*" %n IN ('cs fetch -p org.scala-lang:scala3-library_3:3.0.0') DO java -cp %n:. myMain
```

**Comment by bjornregnell** [comment]
Many thanks @dwijnand ! (I obviously should have persisted a bit more in my impatient googling). Will try it as soon as I get hold of a Windows box, and report back if that long-liner works.

**Comment by sazid** [comment]
On mac/linux, I've setup an alias that can be used easily to run the main method:

Goes into `~/.bashrc`, `~/.zshrc` or any other shell config:
```sh
alias scala3="cs launch scala3-repl:3+ -M dotty.tools.MainGenericRunner -- "
```

Code `MyFile.scala`:
```scala
case class Node(data: Int, children: List[Node])

@main def main =
  val root = Node(10, List())
  println(root)
```

Compile and run the code like so:
```sh
$ scala3-compiler MyFile.scala
$ scala3 main
Node(10,List())
```

**Comment by sjrd** [comment]
With the latest preview build of coursier ([here](https://github.com/coursier/coursier/releases/tag/v2.0.16-200-ge888c6dea)), and with the proposed changes to coursier/apps in https://github.com/coursier/apps/pull/113, I could successfully execute the following steps (on Linux at least):
```bash
~/opt/cs-next$ ls
cs-x86_64-pc-linux
```
The above is the executable downloaded from the preview build assets.

The following command is a fancy invocation of `setup` to use the custom `coursier-apps` and a custom installation directory. Once everything is merged and published, regular users will only do `./cs-x86_64-pc-linux setup`.
```bash
~/opt/cs-next$ ./cs-x86_64-pc-linux setup --install-dir ~/opt/cs-next/install/ --default-channels=false --channel /localhome/doeraene/projects/coursier-apps/apps/resources 
Checking if a JVM is installed
Found a JVM installed under /usr/lib/jvm/java-8-openjdk-amd64.

Checking if ~/opt/cs-next/install is in PATH
  Should we add ~/opt/cs-next/install to your PATH via ~/.profile, ~/.bash_profile? [Y/n] n

Checking if the standard Scala applications are installed
  Installed ammonite
  Installed cs
  Installed coursier
  Installed scala
  Installed scalac
  Installed sbt
  Installed sbtn
  Installed scalafmt

```
```bash
~/opt/cs-next$ ./install/scalac -version
Scala compiler version 3.1.0 -- Copyright 2002-2021, LAMP/EPFL
```
(note that Scala 3.1.0 was installed by default)
```bash
~/opt/cs-next$ mkdir test
```
```bash
~/opt/cs-next$ cd test
```
```bash
~/opt/cs-next/test$ cat > MyFile.scala
case class Node(data: Int, children: List[Node])

@main def main =
  val root = Node(10, List())
  println(root)
```
```bash
~/opt/cs-next/test$ ../install/scalac MyFile.scala 
```
```bash
~/opt/cs-next/test$ ../install/scala MyFile.scala 
Node(10,List())
```
```bash
~/opt/cs-next/test$ ../install/scala 
Welcome to Scala 3.1.0 (1.8.0_292, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                                            
scala> 
```

**Comment by sjrd** [comment]
Equivalent steps work on Windows as well :)

**Comment by bjornregnell** [comment]
Is the path also updated on Windows to include `scala` and `scalac`?

**Comment by sjrd** [comment]
Here I specifically answered `n` to updating the path, since I didn't want to mess with my actual setup, so I did not test that. It seems that that would be unrelated to the present issue anyway, wouldn't it?

**Comment by bjornregnell** [comment]
Well, this issue is about getting a runner on path -- not just the ability to invoke the repl and the compiler... (Compare to the features of the Scala 2 .msi or .deb where you get the `scala.bat` or `scala` bash script that can both run your compiled programs and start the repl. It is a bit dull that you can compile your programs but not run them...)

**Comment by sjrd** [comment]
OK, I just tried again, on Windows, and let it add the install dir to my PATH by just hitting [Enter]. It did correctly add the correct install dir to my user PATH. After that, running the above commands without the `..\install\` prefixes successfully completed as well. Is that what you were asking?

**Comment by bjornregnell** [comment]
Well, adding stuff to path is not easy enough for a beginner. And the good old Scala 2 msi just adds the runner to your path. But it is a step forward that there *is* as runner at least.

**Comment by sjrd** [comment]
I don't understand. As I just said, `cs-x86_64-pc-windows.exe setup` *does* add `cs`, `scala`, `scalac`, `sbt` and other commands to the PATH, automatically. If you double-click on the executable or execute it from your browser's downloads, it will also do `setup` by default on Windows. There is no need to add stuff to path manually.

**Comment by bjornregnell** [comment]
Aha, sorry for my confusion. That's great!! Thanks.

**Comment by bishabosha** [comment]
I think this should be able to be closed now

**Comment by bjornregnell** [comment]
Yes, cs setup installs scala, scalac, scaladoc and later scala-cli when https://github.com/coursier/coursier/pull/2415 is released. :heart:
